drivers: tmp_108: ti tmp108 support

catch-twenty-two opened the pull request:
This adds support for the TI TMP108 temperature sensor. This includes over/under temp interrupt support as well as one shot, continuous conversion and power down modes.

Signed-off-by: Jimmy Johnson <catch22@fastmail.net>

MaureenHelm commented on drivers/sensor/CMakeLists.txt:
```diff
@@ -86,6 +86,7 @@ add_subdirectory_ifdef(CONFIG_STTS751		stts751)
 add_subdirectory_ifdef(CONFIG_SX9500		sx9500)
 add_subdirectory_ifdef(CONFIG_TH02		    th02)
 add_subdirectory_ifdef(CONFIG_TMP007		tmp007)
+add_subdirectory_ifdef(CONFIG_TMP108        tmp108)
```
Please replace spaces with tabs to be consistent with the rest of the file

catch-twenty-two replied:
Ok, Should be done now

MaureenHelm commented on drivers/sensor/Kconfig:
```diff
@@ -212,6 +212,8 @@ source "drivers/sensor/tmp007/Kconfig"
 
 source "drivers/sensor/tmp112/Kconfig"
 
+source "drivers/sensor/tmp108/Kconfig"
```
Please move this between tmp007 and tmp112

MaureenHelm commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,350 @@
+{
+	struct tmp108_data *drv_data = dev->data;
+
+	if (i2c_burst_read(drv_data->i2c_dev,
+			   TMP108_I2C_ADDRESS,
```
Store the I2C address in a const device configuration struct to enable multiple instances of this driver. You can also use `struct i2c_dt_spec` and associated helper functions like `i2c_burst_read_dt`. See #37359 for details

MaureenHelm commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,350 @@
+		return -EINVAL;
+	}
+
+	gpio_pin_configure(drv_data->gpio, DT_INST_GPIO_PIN(0, alert_gpios), GPIO_INPUT);
```
Similar to i2c, you can use `struct gpio_dt_spec` and associated helper functions.

MaureenHelm commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,350 @@
+	return 0;
+}
+
+static struct tmp108_data tmp108_prv_data;
+
+DEVICE_DT_INST_DEFINE(0,
+		      tmp108_init,
+		      NULL,
+		      &tmp108_prv_data,
+		      NULL,
+		      POST_KERNEL,
+		      CONFIG_SENSOR_INIT_PRIORITY,
+		      &tmp108_driver_api);
```
Please update to support multiple instances by using `DT_INST_FOREACH_STATUS_OKAY`. See https://docs.zephyrproject.org/latest/guides/dts/howtos.html#write-device-drivers-using-devicetree-apis for details.

MaureenHelm commented on drivers/sensor/tmp108/tmp108.h:
```diff
@@ -0,0 +1,89 @@
+#define TI_TMP108_REG_LOW_LIMIT     0x02                ///< Low alert set register
+#define TI_TMP108_REG_HIGH_LIMIT    0x03                ///< High alert set register
+
+#define TI_TMP108_CONF_M0           0x0100              ///< Mode 1 configuration register
+#define TI_TMP108_CONF_M1           0x0200              ///< Mode 2 configuration register
+#define TI_TMP108_CONF_CR0          0x2000              ///< Conversion rate 1 configuration register
+#define TI_TMP108_CONF_CR1          0x4000              ///< Conversion rate 2 configuration register
+#define TI_TMP108_CONF_POL          0x0080              ///< Alert pin Polarity register
+#define TI_TMP108_CONF_TM           0x0400
+#define TI_TMP108_CONF_HYS1         0x0020
+#define TI_TMP108_CONF_HYS0         0x0010
```
`BIT(n)` would be useful here. From `#include <sys/util.h>`

MaureenHelm commented on samples/sensor/tmp108/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
```
Please move the new sample to a separate commit

MaureenHelm commented on samples/sensor/tmp108/README.txt:
```diff
@@ -0,0 +1,25 @@
+Title: Thermometer
```
This file should be in .RST format

MaureenHelm reviewed (dismissed):
Please add this sensor to `tests/drivers/build_all/sensor` and fix the failing checks (compliance, scancode, buildkite). 

Thank you for the contribution!

MaureenHelm reviewed (commented):
> For my sample code for the tmp108, should I submit that after this pr is okayed and pulled in? Not sure how to do that separately from the driver without having it fail the tests.

You can push the second commit to the same branch, which will update your PR to contain two commits (one for the driver and one for the sample application)

> Please add this sensor to tests/drivers/build_all/sensor and fix the failing checks (compliance, scancode, buildkite).

`tests/drivers/build_all/sensor/prj.conf` needs to be updated in addition to the dts, and the failing checks still need to be addressed.

catch-twenty-two commented:
@MaureenHelm @mbolivar-nordic thanks for the reviews! Hopefully I have addressed everything as expected. For some reason my commits came in backwards here (driver after sample) where as in my repository they are shown in the opposite direction.

catch-twenty-two commented on drivers/sensor/tmp108/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+
+zephyr_library()
+
+zephyr_library_sources_ifdef(CONFIG_TMP108 tmp108.c tmp108_trigger.c)
```
Done

mbolivar-nordic commented on samples/sensor/tmp108/README.rst:
```diff
@@ -0,0 +1,25 @@
+Title: Thermometer
```
This page's content doesn't fit the usual template and doesn't build properly into HTML docs. Please revise.

You might look at samples/sensor/dht/README.rst for an example of how this is done elsewhere -- although it's up to @MaureenHelm as sensor maintainer to declare what the right template is for sensor sample documentation.

See https://docs.zephyrproject.org/latest/guides/docs/index.html#running-the-documentation-processors for instructions on building a docs change into HTML locally.

mbolivar-nordic commented on samples/sensor/tmp108/src/main.c:
```diff
@@ -0,0 +1,181 @@
+
+	tmp108 = DEVICE_DT_GET_ANY(ti_tmp108);
+
+	if (!tmp108) {
```
This is not enough; you need to check if the sensor device was initialized correctly using `device_is_ready` here too.

mbolivar-nordic reviewed (dismissed):
Thanks for the updates. I noticed a couple of additional things on a second pass; see inline comments.

The commit logs also need some changes to fit with the project policy:

1. the convention is for the shortlog to say `subsystem: the-change-made-to-the-subsystem` (this is one reason why the compliance check is failing)
2. zephyr requires at least one additional line of detail beyond the shortlog and the signed-off-by (this is another)

Beyond that,

3. typos, whitespace, and line length checks are failing (see compliance)

For 1., instead of `Adding ti tmp108 sensor support`, use something like `sensor: add ti tmp108 driver`. And instead of `adding ti tmp108 sensor sample`, do something like `samples: add ti tmp108 sample`.

(If unsure about the `subsystem:` prefix, just run `git log` on the directory or files you are editing and follow along with the crowd.)

For 2., please add some extra context about each patch.

This is what the failed compliance check in the CI results below is complaining about when it says:

```
1: UC3 Title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "adding ti tmp108 sensor sample"
1: UC6 Body has no content, should at least have 1 line.
```

Please address these too.

catch-twenty-two commented:
@mbolivar-nordic thanks, I have applied the updates you suggested.

catch-twenty-two commented:
> Thanks for the updates. I noticed a couple of additional things on a second pass; see inline comments.
> 
> The commit logs also need some changes to fit with the project policy:
> 
> 1. the convention is for the shortlog to say `subsystem: the-change-made-to-the-subsystem` (this is one reason why the compliance check is failing)
> 2. zephyr requires at least one additional line of detail beyond the shortlog and the signed-off-by (this is another)
> 
> Beyond that,
> 
> 1. typos, whitespace, and line length checks are failing (see compliance)
> 
> For 1., instead of `Adding ti tmp108 sensor support`, use something like `sensor: add ti tmp108 driver`. And instead of `adding ti tmp108 sensor sample`, do something like `samples: add ti tmp108 sample`.
> 
> (If unsure about the `subsystem:` prefix, just run `git log` on the directory or files you are editing and follow along with the crowd.)
> 
> For 2., please add some extra context about each patch.
> 
> This is what the failed compliance check in the CI results below is complaining about when it says:
> 
> ```
> 1: UC3 Title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "adding ti tmp108 sensor sample"
> 1: UC6 Body has no content, should at least have 1 line.
> ```
> 
> Please address these too.

Addressed this, thanks!

catch-twenty-two commented:
@MaureenHelm or @mbolivar it looks like all the tests are finally passing, what are next steps?

cfriedt commented:
@MaureenHelm - PTAL when you can

gmarull reviewed (changes requested):
There are a few common issues that should be addressed (style, error propagation), as well as some other questions I raised.

MaureenHelm reviewed (changes requested):
Agree with @gmarull 's comments

MaureenHelm reviewed (changes requested):
Some of the changes introducing the sample application slipped into the first commit (`drivers: sensor: add ti tmp108 driver support`). Please move them to the second commit (`sample: sensor: tmp108: Add ti tmp108 sample`)

catch-twenty-two committed:
drivers: sensor: add ti tmp108 driver support

Adding support for the TI TMP108 temperature sensor. This includes
over/under temp interrupt support as well as one shot, continuous
conversion and power down modes.

Signed-off-by: Jimmy Johnson <catch22@fastmail.net>

catch-twenty-two committed:
sample: sensor: tmp108: Add ti tmp108 sample

This sample writes the tmp108 temperature to the console
once every 3 seconds. There are macro definitions included
for turning off and on alerts if that is set up, and
also using low power one shot mode.

Signed-off-by: Jimmy Johnson <catch22@fastmail.net>

carlescufi commented:
@gmarull @MaureenHelm could you take another look?

mbolivar-nordic commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,406 @@
+	static struct tmp108_data tmp108_prv_data_##inst;		   \
+	static const struct tmp108_config tmp108_config_##inst = {	 \
+		.i2c_spec = I2C_DT_SPEC_INST_GET(inst),			   \
+		.alert_gpio = GPIO_DT_SPEC_INST_GET(inst, alert_gpios) \
```
Since alert-gpios is an optional property, you need to use GPIO_DT_SPEC_INST_GET_OR here instead.

mbolivar-nordic reviewed (commented):
Thanks, I confirm that my existing DT and driver model comments are all addressed. I found another bug, otherwise this LGTM from that perspective.

catch-twenty-two commented:
@mbolivar-nordic or @MaureenHelm Is there a recommended way to rollback the sample changes to get to the previous (driver) commit, change that, then reapply the sample changes? It's a lot of work  to make specific changes to completely separate commits on the same branch, especially since I have had to do this multiple times.

mbolivar-nordic commented:
>  Is there a recommended way to rollback the sample changes to get to the previous (driver) commit, change that, then reapply and make changes to the sample?

Use an interactive git rebase.

mbolivar-nordic commented:
See e.g. step 15 in https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow, although that example is more complicated than I would like.

A simpler example:

```
git checkout add-ti-tmp108-support
git rebase -i HEAD~~
# change the driver commit from 'pick' to 'edit' in the rebase TODO file, save and exit
# edit the driver, and commit it
git rebase --continue
```

catch-twenty-two commented:
> See e.g. step 15 in https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow, although that example is more complicated than I would like.
> 
> A simpler example:
> 
> ```
> git checkout add-ti-tmp108-support
> git rebase -i HEAD~~
> # change the driver commit from 'pick' to 'edit' in the rebase TODO file, save and exit
> # edit the driver, and commit it
> git rebase --continue
> ```

Thanks! Ill give that a shot.

gmarull commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,406 @@
+
+#define TMP108_WAKEUP_TIME_IN_MS 30
+
+#define LOG_LEVEL CONFIG_SENSOR_LOG_LEVEL
```
unused define

gmarull commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,406 @@
+
+#define LOG_LEVEL CONFIG_SENSOR_LOG_LEVEL
+
+LOG_MODULE_REGISTER(TMP108, CONFIG_SENSOR_LOG_LEVEL);
```
by convention: please move this to the top, after includes

gmarull commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,406 @@
+#include "tmp108.h"
+
+/** TI conversion scale from 16 bit int temp value to float */
+
+#define TMP108_TEMP_MULTIPLIER	  62500
+
+/** TMP typical conversion time of 27 ms after waking from sleep */
+
+#define TMP108_WAKEUP_TIME_IN_MS 30
```
nit: unneeded blank lines

gmarull commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,406 @@
+int tmp108_reg_read(const struct device *dev, uint8_t reg, uint16_t *val)
+{
+	const struct tmp108_config *cfg = dev->config;
+	int result = 0;
```
redundant 0 initialization

gmarull commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,406 @@
+	int result = 0;
+
+	result = i2c_burst_read_dt(&cfg->i2c_spec, reg, (uint8_t *) val, 2);
+
```
unneded blank line (and others)

gmarull commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,360 @@
+{
+	const struct tmp108_config *cfg = dev->config;
+
+	uint8_t tx_buf[3] = { reg, val >> 8, val & 0xFF };
```
please, check `byteorder.h` API

catch-twenty-two replied:
please see my above comment.

gmarull replied:
```c
tx_buf[0] = reg;
sys_put_be16(val, &tx_buf[1]);
```

gmarull commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,406 @@
+}
+
+static int tmp108_sample_fetch(const struct device *dev,
+							enum sensor_channel chan)
```
please, validate supported channels here (SENSOR_CHAN_ALL and SENSOR_CHAN_AMBIENT_TEMP)

catch-twenty-two replied:
I'm still not sure when sensor channel all should be used, can you point me to some documentation?

gmarull replied:
see for example https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/lm77/lm77.c#L272

ref. https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/sensor.h#L513

gmarull commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,406 @@
+	int32_t uval;
+
+	if (chan != SENSOR_CHAN_AMBIENT_TEMP &&
+		chan != SENSOR_CHAN_ALL) {
```
SENSOR_CHAN_ALL should not be validated here

gmarull commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,406 @@
+		return -ENOTSUP;
+	}
+
+	uval = (int32_t)(drv_data->sample  >> 4) * TMP108_TEMP_MULTIPLIER;
```
nit: add `U` suffix to all unsigned constants `>> 4U`

gmarull commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,406 @@
+	int result = 0;
+
+	if (chan != SENSOR_CHAN_AMBIENT_TEMP &&
+		chan != SENSOR_CHAN_ALL) {
```
fits into single line
```
	if (chan != SENSOR_CHAN_AMBIENT_TEMP && chan != SENSOR_CHAN_ALL) {
		return -ENOTSUP;
	}
```

gmarull commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,406 @@
+		return -ENOTSUP;
+	}
+
+	switch ((int) attr) {
```
why `int` cast here?

catch-twenty-two replied:
I'm not sure, this might have been left over from testing.

catch-twenty-two replied:
since this includes public attributes that are not listed in the sensor attr enum, I am getting warnings, if I don't cast this as an int example:

```
/home/jimmy/zephyrupstream/zephyr/drivers/sensor/tmp108/tmp108.c:271:2: warning: case value '15' not in enumerated type 'enum sensor_attribute' [-Wswitch]
  271 |  case SENSOR_ATTR_SHUTDOWN_MODE:

```

gmarull commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,406 @@
+	uint16_t reg_value = 0;
+	int result = 0;
+
+	if (chan != SENSOR_CHAN_AMBIENT_TEMP &&
+		chan != SENSOR_CHAN_ALL) {
+		return -ENOTSUP;
+	}
```
fits into single line
```
	if (chan != SENSOR_CHAN_AMBIENT_TEMP && chan != SENSOR_CHAN_ALL) {
		return -ENOTSUP;
	}
```

gmarull commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,406 @@
+		return result;
+	}
+
+	return 0;
```
this function will return 0 even if attr is not supported

gmarull commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,406 @@
+	struct tmp108_data *drv_data = dev->data;
+	const struct tmp108_config *config = dev->config;
+	const struct gpio_dt_spec *alert_gpio = &config->alert_gpio;
+	int result = 0;
```
redundant 0 init

gmarull commented on drivers/sensor/tmp108/tmp108.c:
```diff
@@ -0,0 +1,406 @@
+	return 0;
+}
+
+#define TMP108_INST(inst)						   \
```
`TMP108_DEFINE`?

catch-twenty-two replied:
I'm not sure what you are referring to here?

gmarull replied:
I was suggesting a name change, `INST` doesn't describe the purpose of the macro

gmarull commented on include/drivers/sensor/tmp108.h:
```diff
@@ -0,0 +1,43 @@
+};
+
+/** a mask for the over temp alert bit in the status word*/
+#define OVER_TEMP_MASK 0x1000
+
+/** a mask for the under temp alert bit in the status word*/
+#define UNDER_TEMP_MASK	0x0800
```
`U` suffix

catch-twenty-two commented:
Hi @gmarull thanks for the input, I was able to make the changes requested, I just had question above.

gmarull reviewed (changes requested):
Please check my comments, you may need to review your editor settings

catch-twenty-two commented on drivers/sensor/tmp108/tmp108.h:
```diff
@@ -0,0 +1,88 @@
+#define TI_TMP108_HYSTER_1_C		TI_TMP108_CONF_HYS0
+#define TI_TMP108_HYSTER_2_C		TI_TMP108_CONF_HYS1
+#define TI_TMP108_HYSTER_4_C		(TI_TMP108_CONF_HYS1 | TI_TMP108_CONF_HYS0)
+#define TI_TMP108_HYSTER_MASK	   ~(TI_TMP108_CONF_HYS1 | TI_TMP108_CONF_HYS0)
```
I ran uncrusitfy and tried to fix up the inconsistencies.

gmarull commented on include/drivers/sensor/tmp108.h:
```diff
@@ -0,0 +1,43 @@
+
+enum sensor_attribute_tmp_108 {
+	/** Turn on power saving/one shot mode */
+	SENSOR_ATTR_ONE_SHOT_MODE = SENSOR_ATTR_PRIV_START,
+	/** Shutdown the sensor */
+	SENSOR_ATTR_SHUTDOWN_MODE,
+	/** Turn on continuous conversion */
+	SENSOR_ATTR_CONTINUOUS_CONVERSION_MODE,
+	/** Set the alert pin polarity */
+	SENSOR_ATTR_ALERT_POLARITY
```
please, namespace these custom attributes (see e.g. sgp40)

gmarull commented on samples/sensor/tmp108/src/main.c:
```diff
@@ -0,0 +1,187 @@
+ */
+#define ENABLE_ONE_SHOT             0
+
+/* Set this to 1 to enable alerts for high and low temperatures
+ * although this will work with one shot enabled, it requires
+ * continuous monitoring mode to be enabled to work in real time.
+ * See spec sheet for more details.
+ */
+#define ENABLE_TEMP_ALERT           0
```
use Kconfig for this, e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/sensor/sgp40_sht4x/Kconfig

gmarull commented on samples/sensor/tmp108/src/main.c:
```diff
@@ -0,0 +1,187 @@
+				    SENSOR_CHAN_AMBIENT_TEMP,
+				    &temp_value);
+
+	if (result) {
+		printf("error: sensor_channel_get failed: %d\n", result);
+	}
+
+	printf("temperature is %gC\n", sensor_value_to_double(&temp_value));
```
second printf will be executen, even on error

gmarull reviewed (changes requested):
There are style issues that have not been fixed

catch-twenty-two commented:
> There are style issues that have not been fixed

Is there a style guide for this? I would have thought Uncrusitfy paired with the checkpatch perl script would have sufficed,  if there isn't I would suggest that the Uncrustify script be updated. Manually changing a style to conform to every nuance is very difficult and subjective since it can vary greatly from project to project and person to person. (e.g. a persons preference for grouping functionality in a routine with spacing between calls)

gmarull commented:
> > There are style issues that have not been fixed
> 
> Is there a style guide for this? I would have thought Uncrusitfy paired with the checkpatch perl script would have sufficed, if there isn't I would suggest that the Uncrustify script be updated. Manually changing a style to conform to every nuance is very difficult and subjective since it can vary greatly from project to project and person to person. (e.g. a persons preference for grouping functionality in a routine with spacing between calls)

https://docs.zephyrproject.org/latest/contribute/index.html#coding-style
https://www.kernel.org/doc/html/latest/process/coding-style.html

teburd commented on dts/bindings/sensor/ti,tmp108.yaml:
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2019, Linaro Limited
```
Is this really a linaro created file?

catch-twenty-two replied:
fixed

teburd reviewed (approved):
Looks good to me. I'd honestly suggest filing an issue on uncrustify and the docs here if you ran it and it still didn't conform. The docs suggest using it, I've done just that as well, and yet folks still note that it doesn't conform to the style guide which is I agree a confusing thing to run into. Especially as a new contributor.

catch-twenty-two commented:
> Looks good to me. I'd honestly suggest filing an issue on uncrustify and the docs here if you ran it and it still didn't conform. The docs suggest using it, I've done just that as well, and yet folks still note that it doesn't conform to the style guide which is I agree a confusing thing to run into. Especially as a new contributor.

Agreed, we are preparing to submit more drivers and updates to drivers, I'd hate for things to blocked because of stylistic issues that can't be address without multiple updates to prs.

gmarull commented:
> > Looks good to me. I'd honestly suggest filing an issue on uncrustify and the docs here if you ran it and it still didn't conform. The docs suggest using it, I've done just that as well, and yet folks still note that it doesn't conform to the style guide which is I agree a confusing thing to run into. Especially as a new contributor.
> 
> Agreed, we are preparing to submit more drivers and updates to drivers, I'd hate for things to blocked because of stylistic issues that can't be address without multiple updates to prs.

I agree that it's annoying, and unfortunately tooling like uncrustify seem to be outdated or not working as expected (there are some compliance issues that are likely related to that). You may try with clang-format, from my experience it is not perfect but does a decent job.

MaureenHelm commented:
@gmarull have your requested changes been addressed?

gmarull commented:
> @gmarull have your requested changes been addressed?

I don't see any changes related to my last comments, @catch-twenty-two can you please check?

catch-twenty-two commented:
I haven't had time to look into clang and haven't used that before so I'll probably wont get around to it before the new year. I believe the only problems left are style issues.

catch-twenty-two commented:
@gmarull i took another quick look today, it seems like uncrusitfy has matched the formatting of the other files in the project which do not strictly use only spaces but a mixture with 8 width tabs preceding spaces to align. @MaureenHelm fyi

gmarull reviewed (approved):
I guess it's ok to move on. I'll try to spend some time digging into the status of formatters and see if they can be improved.

catch-twenty-two commented:
> I guess it's ok to move on. I'll try to spend some time digging into the status of formatters and see if they can be improved.

Thanks, just fyi I have opened an issue for this here: https://github.com/zephyrproject-rtos/zephyr/issues/41281
